sdk/core/System.Net.ClientModel/src/Message/RequestBody.cs

```diff
@@ -0,0 +1,181 @@
+
+            public ModelMessageBody(IModel<object> model, ModelReaderWriterOptions options)
+            {
+                _model = model;
+                _options = options;
+                _wireFormat = model.GetWireFormat(options);
+            }
+
+            private ModelWriter Writer
+            {
+                get
+                {
+                    if (_model is not IJsonModel<object> jsonModel)
+                    {
+                        throw new InvalidOperationException("Cannot use Writer with non-IJsonModel model type.");
+                    }
+
+                    _writer ??= new ModelWriter(jsonModel, _options);
+                    return _writer;
+                }
+            }
+
+            private BinaryData Data
+            {
+                get
+                {
+                    if (_model is IJsonModel<object> && _wireFormat == ModelReaderWriterFormat.Json)
+                    {
+                        throw new InvalidOperationException("Should use ModelWriter instead of _model.Write with IJsonModel.");
+                    }
+
+                    _data ??= _model.Write(_options);
+                    return _data;
+                }
+            }
+
+            public override bool TryComputeLength(out long length)
+            {
+                if (_model is IJsonModel<object> && _wireFormat == ModelReaderWriterFormat.Json)
+                {
+                    return Writer.TryComputeLength(out length);
+                }
+
+                length = Data.ToMemory().Length;
+                return true;
+            }
+
+#if NETFRAMEWORK || NETSTANDARD2_0
+            private byte[]? _bytes;
+            private byte[] Bytes => _bytes ??= Data.ToArray();
+#endif
+
+            public override void WriteTo(Stream stream, CancellationToken cancellation)
+            {
+                if (_model is IJsonModel<object> && _wireFormat == ModelReaderWriterFormat.Json)
+                {
+                    Writer.CopyTo(stream, cancellation);
+                    return;
+                }
+
+#if NETFRAMEWORK || NETSTANDARD2_0
+                stream.Write(Bytes, 0, Bytes.Length);
+#else
+                stream.Write(Data.ToMemory().Span);
+#endif
+            }
+
+            public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+            {
+                if (_model is IJsonModel<object> && _wireFormat == ModelReaderWriterFormat.Json)
+                {
+                    await Writer.CopyToAsync(stream, cancellation).ConfigureAwait(false);
+                    return;
+                }
+
+                await stream.WriteAsync(Data.ToMemory(), cancellation).ConfigureAwait(false);
+            }
+
+            public override void Dispose()
+            {
+                var writer = _writer;
+                if (writer != null)
+                {
+                    _writer = null;
+                    writer.Dispose();
```

Comment:
If we can find a way to make this not disposable, then RequestBody doesn't need to be disposable.